101-webapp-for-containers/azuredeploy.json

```diff
@@ -0,0 +1,130 @@
+      }
+    },
+    "alwaysOn": {
+      "type": "bool",
+      "defaultValue": false,
+	  "metadata": {
+        "description": "Whether to ensure the web app gets loaded all the time, rather than being unloaded after been idle. Recommended when you have continuous web jobs running."
+      }
+    },
+    "existingContainerRegistryName": {
+      "type": "string",
+	  "metadata": {
+        "description": "The name of the container registry."
+      }
+    },
+    "dockerImageName": {
+      "type": "string",
+	  "metadata": {
+        "description": "The container image name from the container registry."
+      }
+    },
+    "dockerTag": {
+      "type": "string",
+	  "defaultValue": "latest",
+	  "metadata": {
+        "description": "The container image tag."
+      }
+    },
+    "containerRegistryUsername": {
+      "type": "string",
+	  "metadata": {
+        "description": "The container registry server username."
+      }
+    },
+    "containerRegistryPassword": {
+      "type": "string",
+	  "metadata": {
+        "description": "The container registry server password."
+      }
+    },
+    "appSettings": {
+      "type": "array",
+      "defaultValue": [],
+	  "metadata": {
+        "description": "The web app settings."
+      }
+    }
+  },
+  "resources": [
+    {
+      "name": "[parameters('appServicePlanName')]",
+      "type": "Microsoft.Web/serverfarms",
+      "apiVersion": "2020-06-01",
+      "kind": "linux",
+      "location": "[parameters('location')]",
+      "properties": {
+        "reserved": true
+      },
+      "sku": {
+        "name": "[parameters('sku')]",
+        "capacity": "[parameters('skuCapacity')]"
+      }
+    },
+    {
+      "name": "[parameters('webAppName')]",
+      "type": "Microsoft.Web/sites",
+      "apiVersion": "2020-06-01",
+      "kind": "app,linux,container",
+      "location": "[parameters('location')]",
+      "properties": {
+        "name": "[parameters('webAppName')]",
+        "serverFarmId": "[resourceId('Microsoft.Web/serverfarms', parameters('appServicePlanName'))]",
+        "siteConfig": {
+          "linuxFxVersion": "[concat('DOCKER|', parameters('existingContainerRegistryName'), '.azurecr.io/', parameters('dockerImageName'), ':', parameters('dockerTag'))]",
+          "webSocketsEnabled": "[parameters('webSocketsEnabled')]",
+          "alwaysOn": "[parameters('alwaysOn')]",
+          "appSettings": "[union(parameters('appSettings'), createArray(json('{\"name\": \"WEBSITES_ENABLE_APP_SERVICE_STORAGE\", \"value\": false }'), json(concat('{\"name\": \"DOCKER_REGISTRY_SERVER_URL\", \"value\": \"', concat('https://', parameters('existingContainerRegistryName'), '.azurecr.io/', parameters('dockerImageName')),'\"}')), json(concat('{\"name\": \"DOCKER_REGISTRY_SERVER_USERNAME\", \"value\": \"', parameters('containerRegistryUsername'), '\"}')), json(concat('{\"name\": \"DOCKER_REGISTRY_SERVER_PASSWORD\", \"value\": \"', parameters('containerRegistryPassword'), '\"}'))))]"
```

Comment:
move the createArray code to a variable and just define the array there in JSON (without template functions)... then here you can use the union() function to merge the 2 - will be much more readable.

Reply:
It may be more readable but I can't seem to get the validation procedure to pass if I move the required app settings (as-is) to a variable. I get an `InvalidTemplateDeployment`  with a `ValidationForResourceFailed` and an `AccessToContainerRegistryDenied` code.

Reply:
Can't seem to pass the validation procedure unless I use the template functions. Getting an `InvalidTemplateDeployment` with an access denied error code.

Reply:
go ahead and commit the changes to the PR and I can take a look at the failure...

Reply:
So compared to the last successful commit, you want me to remove the use of the `json` template function from the variable? I will still need to use `concat` to concatenate the variables? Or what string are you suggesting to be passed to `createArray`?

Reply:
just create the appSettings property in json without the json() function and without the createArray() function, e.g.

```
"appSettings": [
  {
    "name": "foo",
    "value": "[parameters('something')]"
  },
  {
    "name": "foo2",
    "bar": "bar2"
  }
]
```

Reply:
You are supposed to be able to add your own custom appSettings using the `appSettings` parameter.

These should then be merged with the appSettings in the variable that are always required (`DOCKER_REGISTRY_SERVER_URL`, `DOCKER_REGISTRY_SERVER_USERNAME`, `DOCKER_REGISTRY_SERVER_PASSWORD`) for the WebApp to be able to authenticate to the ACR and spin up the container.

So you can't really define all appSettings in a single JSON variable, can you?